include/kafka/BrokerMetadata.h

```diff
@@ -111,23 +122,44 @@ struct BrokerMetadata {
     std::map<Partition, PartitionInfo>        _partitions;
 };
 
+inline std::vector<BrokerMetadata::Node>
+BrokerMetadata::nodes() const
+{
+    std::vector<Node> ret;
+    for (const auto& nodeInfo: _nodes)
+    {
+        ret.emplace_back(*nodeInfo.second);
```

Comment:
what about we check the validity of the shared_ptr first before dereference?

Reply:
Or we can consider to define _nodes as std::map<Node::Id, Node>? If the shared_ptr is the source of evil here

Reply:
What about we use unordered_map<Node::Id, Node>? In favor of hash lookup.

Reply:
Most of the cases, there would be not many nodes (brokers) in the `map`, while `unordered_map` might cost a bit more in memory.
https://stackoverflow.com/questions/2196995/is-there-any-advantage-of-using-map-over-unordered-map-in-case-of-trivial-keys

Reply:
> Or we can consider to define _nodes as std::map<Node::Id, Node>? If the shared_ptr is the source of evil here

no, `shared_ptr` is not likely to be evil if with care. 😅